src/client/pmix_client_get.c

```diff
@@ -497,64 +335,192 @@ static void getnb_cbfunc(struct pmix_peer_t *pr, pmix_usock_hdr_t *hdr,
+                            cb->key, cb->procs->rank, nm);
+        cb->value_cbfunc(rc, NULL, cb->cbdata);
+        /* protect the data */
+        cb->procs = NULL;
+        cb->key = NULL;
+        cb->info = NULL;
+        PMIX_RELEASE(cb);
+        return;
+    }
+
+  request:
+    /* if we got here, then we don't have the data for this proc. If we
+     * are a server, or we are a client and not connected, then there is
+     * nothing more we can do */
+    if (pmix_globals.server || (!pmix_globals.server && !pmix_globals.connected)) {
+        cb->value_cbfunc(PMIX_ERR_NOT_FOUND, NULL, cb->cbdata);
+        /* protect the data */
+        cb->procs = NULL;
+        cb->key = NULL;
+        cb->info = NULL;
+        PMIX_RELEASE(cb);
+        return;
+    }
+
+    /* we also have to check the user's directives to see if they do not want
+     * us to attempt to retrieve it from the server */
+    for (n=0; n < cb->ninfo; n++) {
+        if (0 == strcmp(cb->info[n].key, PMIX_OPTIONAL) &&
+            cb->info[n].value.data.flag) {
+            /* they don't want us to try and retrieve it */
+            pmix_output_verbose(2, pmix_globals.debug_output,
+                                "PMIx_Get key=%s for rank = %d, namespace = %s was not found - request was optional",
+                                cb->key, cb->procs->rank, nm);
+            cb->value_cbfunc(PMIX_ERR_NOT_FOUND, NULL, cb->cbdata);
+            /* protect the data */
+            cb->procs = NULL;
+            cb->key = NULL;
+            cb->info = NULL;
+            PMIX_RELEASE(cb);
+            return;
+        }
+    }
 
     /* see if we already have a request in place with the server for data from
      * this nspace:rank. If we do, then no need to ask again as the
      * request will return _all_ data from that proc */
-    PMIX_LIST_FOREACH(cb, &pmix_client_globals.pending_requests, pmix_cb_t) {
-        if (0 == strncmp(cb->nspace, data->cb->nspace, PMIX_MAX_NSLEN) && cb->rank == data->cb->rank) {
+    PMIX_LIST_FOREACH(cbret, &pmix_client_globals.pending_requests, pmix_cb_t) {
+        if (0 == strncmp(cbret->nspace, nm, PMIX_MAX_NSLEN) &&
+            cbret->rank == cb->rank) {
             /* we do have a pending request, but we still need to track this
              * outstanding request so we can satisfy it once the data is returned */
-            pmix_list_append(&pmix_client_globals.pending_requests, &data->cb->super);
-            goto exit;
+            pmix_list_append(&pmix_client_globals.pending_requests, &cb->super);
+            return;
         }
     }
 
     /* we don't have a pending request, so let's create one - don't worry
      * about packing the key as we return everything from that proc */
-    msg = pack_get(data->cb->nspace, data->cb->rank, data->info, data->ninfo, PMIX_GETNB_CMD);
+    msg = pack_get(nm, cb->rank, cb->info, cb->ninfo, PMIX_GETNB_CMD);
```

Comment:
@rhc54  and here cb->rank is used, which we haven't initialized because was relying on cb->procs.